### PR TITLE
adapt UCTE helpers depending on element sought (contingency vs. cnec vs. topo vs pst) + add Hvdc UCTE helper

### PR DESCRIPTION
**Pull request by baptsegu**

### CODE REVIEW

**Review by phiedw — CHANGES_REQUESTED** [review]

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java [hunk 1]
```diff
@@ -90,42 +103,86 @@ public double getCurrentLimit(Branch.Side side) {
...
+        if (Objects.isNull(cnecElement)) {
+            return false;
+        }
+        this.branchIdInNetwork = cnecElement.getId();
+
+        if (cnecElement instanceof TieLine) {
+            checkBranchNominalVoltage((Branch<?>) cnecElement);
+            checkTieLineCurrentLimits((TieLine) cnecElement);
+        }
+        if (cnecElement instanceof Branch) {
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java:131]: it is an else if, thanks :)

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java [hunk 2]
```diff
@@ -90,42 +103,86 @@ public double getCurrentLimit(Branch.Side side) {
...
+        }
+        return true;
+    }
+
+    private boolean interpretAsHalfLine(Network network) {
+        Optional<TieLine> tieLine = network.getBranchStream()
+                .filter(TieLine.class::isInstance)
+                .map(TieLine.class::cast)
+                .filter(b -> b.getHalf1().getId().equals(branchId) || b.getHalf2().getId().equals(branchId))
+                .findAny();
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java:148]: I agree.
This method 'interpretAsHalfLine()' is actually not used yet, i just put it in anticipation of the issue with the CIM Crac importer that we recently identified.

As we mostly work with UCTE networks, it's hard to know whether the impact is big or not. As discussed this morning, I propose that we postpone this question for the Cnec CIM Crac import :angel:.

If you prefer, i can remove this method in between.

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteCnecElementHelper.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+     */
+    public UcteCnecElementHelper(String ucteBranchId, UcteNetworkAnalyzer ucteNetworkAnalyzer) {
+        super(ucteBranchId);
+        if (isValid) {
+            interpretWithNetworkAnalyzer(ucteNetworkAnalyzer);
+        }
+    }
+
+    @Override
+    public String getIdInNetwork() {
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteCnecElementHelper.java:88]: removed

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteCnecElementHelper.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    public boolean isHalfLine() {
+        return isHalfLine;
+    }
+
+    @Override
+    public Branch.Side getHalfLineSide() {
+        return halfLineSide;
+    }
+
+    protected void interpretWithNetworkAnalyzer(UcteNetworkAnalyzer networkHelper) {
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteCnecElementHelper.java:130]: renamed

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteHvdcElementHelper.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    public UcteHvdcElementHelper(String ucteBranchId, UcteNetworkAnalyzer ucteNetworkAnalyzer) {
+        super(ucteBranchId);
+        if (isValid) {
+            interpretWithNetworkAnalyzer(ucteNetworkAnalyzer);
+        }
+    }
+
+    @Override
+    public String getIdInNetwork() {
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteHvdcElementHelper.java:77]: removed

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteHvdcElementHelper.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     * If the branch is valid, returns a boolean indicating whether or not the from/to are
+     * inverted in the network, compared to the values originally used in the constructor
+     * of the UcteBranchHelper
+     */
+    @Override
+    public boolean isInvertedInNetwork() {
+        return isInvertedInNetwork;
+    }
+
+    protected void interpretWithNetworkAnalyzer(UcteNetworkAnalyzer networkHelper) {
```
[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/ucte/UcteHvdcElementHelper.java:91]: renamed

**Review by phiedw — DISMISSED** [review]

File: data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (phiedw) [reply data/crac/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation_util/iidm/IidmCnecElementHelper.java:147]: It's fine for now, we'll see when we work on the CIM import.

**Review by pet-mit — APPROVED** [review]

**Pull request merged**